src/networking/crypto/keytypes.rs

```diff
@@ -0,0 +1,226 @@
+      PublicKey::OpenSSLMedium(_) => openssl::nid::Nid::SECT409K1,
+      PublicKey::OpenSSLHigh(_) => openssl::nid::Nid::SECT571R1,
+      _ => return None,
+    })
+  }
+
+  pub fn to_vec(&self) -> Option<Vec<u8>>{
+    Some(match self{
+      PublicKey::Ed25519(i) => {
+        //translates to "LibNaCLPK:" which is the (very silly) prefix used by py-ipv8
+        let mut res = vec![76, 105, 98, 78, 97, 67, 76, 80, 75, 58];
+        res.extend_from_slice(i.as_ref());
+        res
+      },
+      PublicKey::OpenSSLHigh(i) |
+      PublicKey::OpenSSLMedium(i) |
+      PublicKey::OpenSSLLow(i) |
+      PublicKey::OpenSSLVeryLow(i) => match match openssl::pkey::PKey::from_ec_key(i.to_owned()){
+        Ok(i) => i,
+        Err(_) => return None,
+      }.public_key_to_der(){
+        Ok(i) => i,
+        Err(_) => return None
+      },
+      _ => return None,
+    })
+  }
+
+  /// return signature length in bytes
+  pub fn signature_len(&self) -> Option<usize>{
+    Some(match self{
+      PublicKey::Ed25519(i) => 64, // the length of an ed25519 signature is always 64 bytes
+      PublicKey::OpenSSLHigh(i) |
+      PublicKey::OpenSSLMedium(i) |
+      PublicKey::OpenSSLLow(i) |
+      PublicKey::OpenSSLVeryLow(i) => get_signature_length(self.get_curve()?)? as usize
+    })
+  }
+
+  pub fn from_vec(data: Vec<u8>) -> Option<Self>{
+    // literally "LibNaCLPK:"
+    let ed25519prefix = &[76, 105, 98, 78, 97, 67, 76, 80, 75, 58];
+
+    if data.starts_with(ed25519prefix){
+      // libnacl
+      Some(PublicKey::Ed25519(ed25519::PublicKey::from_slice(&data[ed25519prefix.len()..])?))
+    }else{
+      // openssl DER encoded. Pem always is base64 encoded DER with a header and trailer. py-ipv8 appends these headers and trailers
+      // which is very inefficient. We just keep it as DER as that's basically how we get it from the deserialization process.
+      let pkey = match openssl::pkey::PKey::public_key_from_der(&*data){
+        Ok(i) => i,
+        Err(_) => return None
+      };
+      let eckey = match (*pkey).ec_key(){
+        Ok(i) => i,
+        Err(_) => return None
+      };
+
+      //get the type of key and convert it to a PublicKey enum type
+      Some(match eckey.group().curve_name()?{
+        openssl::nid::Nid::SECT163K1 => PublicKey::OpenSSLVeryLow(eckey),
+        openssl::nid::Nid::SECT233K1 => PublicKey::OpenSSLLow(eckey),
+        openssl::nid::Nid::SECT409K1 => PublicKey::OpenSSLMedium(eckey),
+        openssl::nid::Nid::SECT571R1 => PublicKey::OpenSSLHigh(eckey),
+        _ => return None
+      })
+    }
+  }
+}
+
+pub enum PrivateKey{
+  OpenSSLVeryLow(openssl::ec::EcKey<openssl::pkey::Private>),
+  OpenSSLLow(openssl::ec::EcKey<openssl::pkey::Private>),
+  OpenSSLMedium(openssl::ec::EcKey<openssl::pkey::Private>),
+  OpenSSLHigh(openssl::ec::EcKey<openssl::pkey::Private>),
+  OpenSSLVeryHigh(openssl::ec::EcKey<openssl::pkey::Private>),
+  Ed25519(ed25519::SecretKey),
+}
+
+//impl PartialEq for PrivateKey{
```

Comment:
Can you either delete this or uncomment it?

Reply:
same thing